File: backend/app/utils/skill_loader.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List, Optional
from xml.sax.saxutils import escape

import yaml

from app.core.config import OMNIAGE_ROOT, get_agent_base_dir, get_agent_skill_ids
from app.utils.disabled_skills import DisabledSkills


def _xml_text(s: str) -> str:
    """XML 元素文本转义（&, <, >）。"""
    return escape(s, entities={'"': "&quot;", "'": "&apos;"})


@dataclass(frozen=True)
class SkillHead:
    """SKILL.md 发现阶段：仅 frontmatter + 路径，不含正文。"""

    skill_id: str
    source: str  # "bundled" | "user"
    name: str
    description: str
    skill_md_path: Path


def bundled_skills_dir() -> Path:
    """仓库内官方 skill 根目录：<OMNIAGE_ROOT>/config/agents/skills/<skill_id>/SKILL.md"""
    return OMNIAGE_ROOT / "config" / "agents" / "skills"


def parse_skill_frontmatter(skill_md_path: Path) -> Optional[Dict[str, Any]]:
    content = skill_md_path.read_text(encoding="utf-8")
    if not content.startswith("---"):
        return None
    parts = content.split("---", 2)
    if len(parts) < 3:
        return None
    meta = yaml.safe_load(parts[1])
    if not isinstance(meta, dict):
        return None
    return meta


def read_skill_head(skill_id: str, skill_md_path: Path, source: str) -> Optional[SkillHead]:
    meta = parse_skill_frontmatter(skill_md_path)
    if not meta:
        return None
    name = meta.get("name")
    description = meta.get("description")
    if name is None or description is None:
        return None
    name_s, desc_s = str(name).strip(), str(description).strip()
    if not name_s or not desc_s:
        return None
    return SkillHead(
        skill_id=skill_id,
        source=source,
        name=name_s,
        description=desc_s,
        skill_md_path=skill_md_path,
    )


def load_skill_body(skill_root: Path, skill_id: str) -> str:
    """加载 SKILL.md 正文（去掉 YAML 头）。"""
    skill_file = skill_root / skill_id / "SKILL.md"
    content = skill_file.read_text(encoding="utf-8")
    if content.startswith("---"):
        parts = content.split("---", 2)
        if len(parts) >= 3:
            return parts[2].strip()
    return content.strip()


def load_skill(skill_path: Path, skill_name: str) -> str:
    """兼容旧接口：等价于 load_skill_body。"""
    return load_skill_body(skill_path, skill_name)


def discover_skills_for_agent(agent_id: str, disabled_skills: set[str] | None = None) -> List[SkillHead]:
    """
    发现某 agent 可用 skill：仅解析各 SKILL.md 的 frontmatter。
    顺序：config 中 skills 列表顺序；再追加用户目录下（仅不在该列表中的 skill_id，字母序）。
    同名 skill_id：用户目录覆盖仓库内条目。
    disabled_skills 中的 skill_id 会被跳过（不加入返回列表）。
    """
    disabled = disabled_skills or set()
    bundled_root = bundled_skills_dir()
    merged: Dict[str, SkillHead] = {}
    ordered: List[str] = []

    for skill_id in get_agent_skill_ids(agent_id):
        if skill_id in disabled:
            continue
        p = bundled_root / skill_id / "SKILL.md"
        if not p.is_file():
            continue
        head = read_skill_head(skill_id, p, "bundled")
        if head:
            merged[skill_id] = head
            ordered.append(skill_id)

    user_root = get_agent_base_dir(agent_id) / ".agent" / "skills"
    user_only: List[str] = []
    if user_root.is_dir():
        for sub in sorted(user_root.iterdir()):
            if not sub.is_dir():
                continue
            sid = sub.name
            if sid in disabled:
                continue
            sm = sub / "SKILL.md"
            if not sm.is_file():
                continue
            head = read_skill_head(sid, sm, "user")
            if not head:
                continue
            merged[sid] = head
            if sid not in ordered:
                user_only.append(sid)

    out: List[SkillHead] = []
    for sid in ordered:
        if sid in merged:
            out.append(merged[sid])
    for sid in sorted(user_only):
        out.append(merged[sid])
    return out


def format_skills_discovery_xml(heads: List[SkillHead]) -> str:
    """
    将发现的 SKILL 头信息序列化为一段 XML（供模型解析）。
    每个 skill 含：id、source、name、description、path（SKILL.md 绝对路径）。
    """
    if not heads:
        return ""
    parts = ["<skills>"]
    for h in heads:
        path_str = str(h.skill_md_path.resolve())
        parts.append(
            f'  <skill id="{_xml_text(h.skill_id)}" source="{_xml_text(h.source)}">'
        )
        parts.append(f"    <name>{_xml_text(h.name)}</name>")
        parts.append(f"    <description>{_xml_text(h.description)}</description>")
        parts.append(f"    <path>{_xml_text(path_str)}</path>")
        parts.append("  </skill>")
    parts.append("</skills>")
    return "\n".join(parts)


def discover_skills_xml_for_agent(agent_id: str) -> str:
    """发现某 agent 可用 skill（含 disable 过滤），返回仅含头信息的 XML 字符串；无 skill 时返回空串。"""
    disabled = DisabledSkills.load(agent_id)
    return format_skills_discovery_xml(
        discover_skills_for_agent(agent_id, disabled_skills=disabled.skill_ids)
    )


def prepend_skill_catalog_xml_to_system_prompt(base_system_prompt: str, agent_id: str) -> str:
    """
    将技能目录 XML 动态置于 system prompt 最前；无可用 skill 时等价于 base_system_prompt。
    """
    xml_block = discover_skills_xml_for_agent(agent_id).strip()
    base = (base_system_prompt or "").strip()
    if not xml_block:
        return base_system_prompt if base_system_prompt else ""
    if not base:
        return xml_block
    return f"{xml_block}\n\n{base}"


def format_skill_catalog_lines(heads: List[SkillHead]) -> str:
    """拼成给模型看的技能目录（仅头信息）。"""
    lines = []
    for h in heads:
        lines.append(f"[{h.source}] {h.skill_id} — {h.name}: {h.description}")
    return "\n".join(lines)

File: backend-rs/src/service/skill_loader.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};

use crate::core::config::{get_agent_base_dir, get_agent_config};
use crate::service::disabled_skills::DisabledSkills;
use crate::utils::helpers::normalize_path;

/// Skill 元信息（对应 Python SkillHead）
#[derive(Debug, Clone)]
pub struct SkillHead {
    pub skill_id: String,
    pub name: String,
    pub description: String,
    pub source: String, // "bundled" | "user"
    pub skill_md_path: PathBuf,
}

/// 解析 SKILL.md 的 YAML frontmatter
#[derive(Debug)]
pub struct SkillMeta {
    pub name: String,
    pub description: String,
}

/// 解析 SKILL.md 文件，提取 YAML frontmatter
pub fn parse_skill_md(content: &str) -> Option<SkillMeta> {
    let trimmed = content.trim();
    if !trimmed.starts_with("---") {
        return None;
    }
    let without_start = &trimmed[3..];
    let end = without_start.find("\n---")?;
    let frontmatter_str = &without_start[..end];

    let name = frontmatter_str
        .lines()
        .find_map(|line| line.strip_prefix("name:").or_else(|| line.strip_prefix("name: ")))
        .map(|s| s.trim().trim_matches('"').to_string())
        .unwrap_or_default();

    let description = frontmatter_str
        .lines()
        .find_map(|line| line.strip_prefix("description:").or_else(|| line.strip_prefix("description: ")))
        .map(|s| s.trim().trim_matches('"').to_string())
        .unwrap_or_default();

    if name.is_empty() {
        return None;
    }

    Some(SkillMeta { name, description })
}

fn read_skill_head(skill_id: &str, path: &Path, source: &str) -> Option<SkillHead> {
    let content = std::fs::read_to_string(path).ok()?;
    let meta = parse_skill_md(&content)?;
    Some(SkillHead {
        skill_id: skill_id.to_string(),
        name: meta.name,
        description: meta.description,
        source: source.to_string(),
        skill_md_path: path.to_path_buf(),
    })
}

/// XML 转义
fn xml_text(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&apos;")
}

/// 内置 skill 根路径（统一 config/agents/skills/）
fn bundled_skills_dir() -> PathBuf {
    crate::core::config::get_config_dir()
        .join("agents")
        .join("skills")
}

/// 解析 run_command 在 cwd=skills 时的工作目录（与 Python _resolve_run_cwd 一致）
pub fn resolve_skill_run_dir(workspace: &str, skill_name: &str) -> Result<PathBuf, String> {
    let raw = skill_name.trim();
    if raw.is_empty() {
        return Err("Error: skill_name is required when cwd=skills".to_string());
    }
    let safe_path = PathBuf::from(raw);
    let safe = safe_path
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("");
    if safe.is_empty() || safe == "." || safe == ".." {
        return Err("Error: invalid skill_name".to_string());
    }

    let ws = PathBuf::from(workspace);
    let ws_resolved = ws.canonicalize().unwrap_or(ws);
    let user_root = ws_resolved
        .parent()
        .ok_or_else(|| "Error: cannot resolve parent of workspace".to_string())?
        .join("skills");
    let user_dir = user_root.join(safe);
    if user_dir.is_dir() {
        return Ok(normalize_path(
            user_dir.canonicalize().map_err(|e| format!("Error: {e}"))?,
        ));
    }

    let bundled_dir = bundled_skills_dir().join(safe);
    if bundled_dir.is_dir() {
        return Ok(normalize_path(
            bundled_dir.canonicalize().map_err(|e| format!("Error: {e}"))?,
        ));
    }

    std::fs::create_dir_all(&user_dir).map_err(|e| format!("Error: {e}"))?;
    Ok(normalize_path(
        user_dir.canonicalize().map_err(|e| format!("Error: {e}"))?,
    ))
}

/// 发现指定 agent 可用的所有 skill
pub fn discover_skills_for_agent(agent_id: &str) -> Vec<SkillHead> {
    let disabled = DisabledSkills::load(agent_id);
    let bundled_root = bundled_skills_dir();
    let user_root = get_agent_base_dir(agent_id).join(".agent").join("skills");

    let mut merged: HashMap<String, SkillHead> = HashMap::new();
    let mut ordered: Vec<String> = Vec::new();

    // 1. 从配置中读取有序 skill 列表
    if let Some(cfg) = get_agent_config(agent_id) {
        if let Some(skill_ids) = &cfg.skills {
            for skill_id in skill_ids {
                if disabled.is_disabled(skill_id) {
                    continue;
                }
                let p = bundled_root.join(skill_id).join("SKILL.md");
                if p.is_file() {
                    if let Some(head) = read_skill_head(skill_id, &p, "bundled") {
                        if !merged.contains_key(skill_id) {
                            merged.insert(skill_id.clone(), head);
                            ordered.push(skill_id.clone());
                        }
                    }
                }
            }
        }
    }

    // 2. 扫描用户目录下的 user skill
    if user_root.exists() {
        let mut user_ids: Vec<String> = Vec::new();
        if let Ok(entries) = std::fs::read_dir(&user_root) {
            for entry in entries.flatten() {
                let skill_id = entry.file_name().to_string_lossy().to_string();
                if merged.contains_key(&skill_id) || disabled.is_disabled(&skill_id) {
                    continue;
                }
                let p = user_root.join(&skill_id).join("SKILL.md");
                if p.is_file() {
                    user_ids.push(skill_id);
                }
            }
        }
        user_ids.sort();
        for skill_id in user_ids {
            let p = user_root.join(&skill_id).join("SKILL.md");
            if let Some(head) = read_skill_head(&skill_id, &p, "user") {
                merged.insert(skill_id.clone(), head);
                ordered.push(skill_id);
            }
        }
    }

    ordered.into_iter().filter_map(|id| merged.remove(&id)).collect()
}

/// 发现某 agent 可用 skill（含 disable 过滤），返回 XML 字符串
pub fn discover_skills_xml_for_agent(agent_id: &str) -> String {
    let skills = discover_skills_for_agent(agent_id);
    format_skills_discovery_xml(&skills)
}

/// 将 skill 列表渲染为 XML 目录（与 Python 端 format_skills_discovery_xml 一致）
pub fn format_skills_discovery_xml(skills: &[SkillHead]) -> String {
    if skills.is_empty() {
        return String::new();
    }
    let mut parts: Vec<String> = Vec::new();
    parts.push("<skills>".to_string());
    for s in skills {
        let path_str = normalize_path(s.skill_md_path.canonicalize()
            .unwrap_or_else(|_| s.skill_md_path.clone()))
            .to_string_lossy()
            .to_string();
        parts.push(format!(
            r#"  <skill id="{}" source="{}">"#,
            xml_text(&s.skill_id),
            xml_text(&s.source),
        ));
        parts.push(format!(r#"    <name>{}</name>"#, xml_text(&s.name)));
        parts.push(format!(r#"    <description>{}</description>"#, xml_text(&s.description)));
        parts.push(format!(r#"    <path>{}</path>"#, xml_text(&path_str)));
        parts.push("  </skill>".to_string());
    }
    parts.push("</skills>".to_string());
    parts.join("\n")
}

// ── Tests ────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_skill_md_valid() {
        let content = r#"---
name: test-skill
description: A test skill
---

# Skill body here
"#;
        let meta = parse_skill_md(content).expect("should parse");
        assert_eq!(meta.name, "test-skill");
        assert_eq!(meta.description, "A test skill");
    }

    #[test]
    fn test_parse_skill_md_no_frontmatter() {
        let content = "# Just a markdown file\nno frontmatter here";
        assert!(parse_skill_md(content).is_none());
    }

    #[test]
    fn test_parse_skill_md_missing_name() {
        let content = "---\ndescription: only desc\n---\nbody";
        assert!(parse_skill_md(content).is_none());
    }

    #[test]
    fn test_parse_skill_md_quoted_values() {
        let content = r#"---
name: "quoted-name"
description: "quoted desc"
---
"#;
        let meta = parse_skill_md(content).expect("should parse");
        assert_eq!(meta.name, "quoted-name");
        assert_eq!(meta.description, "quoted desc");
    }

    #[test]
    fn test_format_skills_xml_empty() {
        assert_eq!(format_skills_discovery_xml(&[]), "");
    }

    #[test]
    fn test_format_skills_xml_with_items() {
        let skills = vec![
            SkillHead {
                skill_id: "web-search".into(),
                name: "Web Search".into(),
                description: "Search the web".into(),
                source: "bundled".into(),
                skill_md_path: PathBuf::from("/tmp/skills/web-search/SKILL.md"),
            },
        ];
        let xml = format_skills_discovery_xml(&skills);
        assert!(xml.contains("<skills>"));
        assert!(xml.contains(r#"id="web-search""#));
        assert!(xml.contains(r#"source="bundled""#));
        assert!(xml.contains("<name>Web Search</name>"));
        assert!(xml.contains("<description>Search the web</description>"));
        assert!(xml.contains("<path>"));
        assert!(xml.contains("</skill>"));
        assert!(xml.contains("</skills>"));
    }

    #[test]
    fn test_format_skills_xml_escapes_special_chars() {
        let skills = vec![
            SkillHead {
                skill_id: "a&b".into(),
                name: "A < B".into(),
                description: "Use > and <".into(),
                source: "user".into(),
                skill_md_path: PathBuf::new(),
            },
        ];
        let xml = format_skills_discovery_xml(&skills);
        assert!(xml.contains("a&amp;b"));
        assert!(xml.contains("A &lt; B"));
        assert!(xml.contains("&gt;"));
    }
}
